Feature/7242 rename node same name

RonnyB71 opened the pull request:
<!-- Thank you for contributing to Altinn:) We know this isn't the fun part, but please make sure you follow our [contributing guidelines](../../CONTRIBUTING.md) and put the same effort into the pull request as you did into the code and it should soon find it's way to master. -->

# Avoid sibling nodes with the same name
Fixed so the model editor doesn't allow sibling nodes to be named the same when renaming a node in properties pane.

## Description
@Febakke For your evaluation: There is an inconsistency in how the UI currently handles this when renaming a node and when adding/renaming nodes under the properties pane/fields. On the name property itself this fix will result in an error message saying 'Name already in use', while on the properties pane/fields a unique number is always added even if you try to rename a field with no message saying why you get an arbitrary number added at the end.

## Fixes
- #7242

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green
~~- [ ] Documentation is updated with a separate linked PR in altinn-docs (if applicable)~~

haakemon commented on src/studio/src/designer/frontend/packages/schema-editor/src/components/SchemaInspector.tsx:
```diff
@@ -98,6 +98,22 @@ export const isValidName = (name: string) => {
   return Boolean(name.match(/^[a-zA-ZæÆøØåÅ][a-zA-Z0-9_.\-æÆøØåÅ ]*$/));
 };
 
+export const isNameInUse = (params: { parentSchema: UiSchemaItem | null, path: string, name: string }) => {
```
I would simplify this to just destruct the parameters directly, and remove the `params`. IMO it makes it a bit easier to read, and avoids a lot of repetition - but might just be my preference :)

The checks can also be simplified a bit to avoid the many return statments, which can be confusing I think.

```typescript
const isNameInUse = ({ parentSchema: UiSchemaItem | null, path: string, name: string }) => {
  // Check if the parent node has other children with the same name.
  if (
    parentSchema?.properties.some(
      (p) => p.displayName === name && p.path !== path
    )
  ) {
    return true;
  }
  
  return false;
};
```

lorang92 replied:
I'm just curious here and haven't looked at how the `parentSchema` is fetched. But does this also cover the case for items on root which don't actually have a parent schema?

RonnyB71 replied:
Good point @lorang92 - seems like this is thightly connected to the use of the properties keyword (as everything else in this area) and hence this change doesn't make things worse in that regards, but I'll see if I can improve it to future proof it so we don't have to revisit this when we make that bigger change.

haakemon reviewed (commented):
Would be good to add some unit tests for the `isNameInUse` function :) 

I also added some feedback on the function itself, but that is more preference, and not required changes :)

RonnyB71 commented:
> Would be good to add some unit tests for the `isNameInUse` function :)

I had a look at that the first time, but in order to provoke this error I would need to use both the Editor component and the SchemaInspector component - and then it's not really a unit test. Perhaps a candidate for a cypress test? Happy to show you in case I got it wrong.

> I also added some feedback on the function itself, but that is more preference, and not required changes :)

Implemented.

haakemon commented:
> and then it's not really a unit test

I would disagree - the "unit" here would be the function `isNameInUse`, as its a standalone function.

Having these tests would be very beneficial in the future, if an extension would be made to this function, or if it would be refactored. Then it would be easy to see that you would not break existing functionality.

A cypress test would "go bigger" and test the functionality as a whole (integration test, or end2end test), and would perhaps also be beneficial - but it servers a different purpose IMO :)

RonnyB71 commented:
> > and then it's not really a unit test
> 
> I would disagree - the "unit" here would be the function `isNameInUse`, as its a standalone function.
> 
> Having these tests would be very beneficial in the future, if an extension would be made to this function, or if it would be refactored. Then it would be easy to see that you would not break existing functionality.
> 
> A cypress test would "go bigger" and test the functionality as a whole (integration test, or end2end test), and would perhaps also be beneficial - but it servers a different purpose IMO :)

I don't disagree in any of your points:) I guess it's my lack of knowledge on how to test this in jest in isolation, because last time we looked at it we hit the same wall. I'll ping you so we can have another look:)

haakemon reviewed (approved):
Nice :)
